### PR TITLE
Server Side layer search

### LINKED ISSUE #702: Search Layers show theme link if not in current theme
If layer search result is in current theme, show no link to other theme.

If layer search result is not in current theme, show link to other theme(s)

### LINKED ISSUE #1215: Layersearch: also search in metadata keywords
example: to search for "zones de protection des sources" you have to type ZPS ... 
idem for hq10

### PR DESCRIPTION
**Pull request by petzlux**
fix #1215 
fix #702

### CODE REVIEW

File: geoportailv3/scripts/layers2es.py [hunk 1]
```diff
@@ -0,0 +1,164 @@
...
+
+    layers = []
+    roles = DBSession.query(Role).all()
+    for role in roles:
+        role_layers = []
+        req = Request.blank('/themes?interface=desktop&version=2&role_id='
+                            + str(role.id))
+        req.registry = request.registry
+        req.user = None
+        req.remote_addr = '127.0.0.1'
```
Comment (rmichaelis) [inline geoportailv3/scripts/layers2es.py]: @petzlux if the REMOTE_ADDR header is not set the role_id will be always None see here : https://github.com/camptocamp/c2cgeoportal/blob/233e68fbb130e82a8600725df86ffc06a5e645a5/c2cgeoportal/views/entry.py#L1179-L1180

**Comment by petzlux** [comment]
still todo:

- [ ] merge role ids for given layer

**Comment by petzlux** [comment]
@rmichaelis can you please review ?

**Comment by petzlux** [comment]
Still failing travis due to ngeo build error ....

**Comment by rmichaelis** [comment]
@petzlux Problem solved last week. Try to rebase on master.

File: geoportailv3/static/js/search/searchdirective.js [hunk 2]
```diff
@@ -374,11 +345,11 @@ app.SearchDirectiveController = function($scope, $compile, gettextCatalog,
...
     },
     templates: /** @type {TypeaheadTemplates} */({
-      header: function() {
+      header: goog.bind(function() {
         return '<div class="header">' +
-            gettextCatalog.getString('Background Layers') +
+            this.gettextCatalog_.getString('Background Layers') +
             '</div>';
-      },
+      }, this),
```
Comment (rmichaelis) [inline geoportailv3/static/js/search/searchdirective.js]: In ngeo they use now` function(){...}.bind(this);` instead of `goog.bind(function(){...},this)`

**Comment by petzlux** [comment]
thanks have done ... waiting for travis

File: geoportailv3/static/js/search/searchdirective.js (hunk 2 above)
Comment (petzlux) [inline geoportailv3/static/js/search/searchdirective.js]: okay but then we should migrate all of our codebase to this version. ... another PR !

Comment (rmichaelis) [inline geoportailv3/static/js/search/searchdirective.js]: Of course, no need to modify it. It's only to point it out.

**Comment by rmichaelis** [comment]
I get the following error with layersearch webservice :

>  NotFoundError: NotFoundError(404, u'IndexMissingException[[index_layers] missing]', {u'status': 404, u'error': u'IndexMissingException[[index_layers] missing]'}) (exception showing exception: string indices must be integers)

**Comment by petzlux** [comment]
You need to build an index there is a build target for that

**Comment by petzlux** [comment]
`make -f wepa.mk recreate-search-layers`

**Comment by petzlux** [comment]
okay moved elasticsearch index deletion to just before index operation, should result in virtually no downtime when recreating index

**Comment by rmichaelis** [comment]
Searching for "Courbe de niveau" returns nothing, but a layer "Courbes de niveau" is existing.

File: geoportailv3/views/fulltextsearch.py [hunk 3]
```diff
@@ -123,3 +125,83 @@ def fulltextsearch(self):
...
+            filters['should'].append({"term": {"public": True}})
+            filters['should'].append({"term": {"role_id": role_id}})
+
+        es = get_elasticsearch(self.request)
+        layer_index = get_index(self.request) + '_layers'
+        search = es.search(index=layer_index,
+                           body=query_body,
+                           size=limit*4)
+        objs = search['hits']['hits']
+        print objs
```
Comment (rmichaelis) [inline geoportailv3/views/fulltextsearch.py]: Please remove this

Comment (petzlux) [inline geoportailv3/views/fulltextsearch.py]: ? sorry ? dont understand

Comment (rmichaelis) [inline geoportailv3/views/fulltextsearch.py]: You may want to delete "print objs" that is only for debug purpose

**Comment by rmichaelis** [comment]
geoportailv3-client.mo is never create/updated.
make_localizer needs an updated .mo file

**Comment by rmichaelis** [comment]
@petzlux .mo files are now created. Could you please try, if now all works as expected ?

**Comment by petzlux** [comment]
LGTM ! 

@jaykayone for deploy need to run `make -f {yourbuildfile}.mk recreate-search-layers` to build the layers metadata search index

**Comment by jaykayone** [comment]
can you have a non m-developer eg @adonno test the new functionality for feedback?

if ok, you can merge, @rmichaelis can you adapt update_all.sh accordingly and deploy please?

**Comment by rmichaelis** [comment]
Here to test : http://devv3.geoportail.lu/renaud/

**Comment by jaykayone** [comment]
term base map shows no results

**Comment by petzlux** [comment]
@jaykayone drink another beer and enjoy your holiday :-)

Background Layers are searched on the client, and what we call basemap informally is called Road Map  ....

**Comment by jaykayone** [comment]
any reason why we do not search them on server? 
base map should be keyword in MD

**Comment by rmichaelis** [comment]
Looks good to everyone.
Some new remarks here #1388

I'll deploy today

**Comment by petzlux** [comment]
👍

**Pull request merged**